### ISSUE TITLE
Lock extend should be able to take new timeout value

### ISSUE THREAD

**Issue by thomasst** [andymccurdy/redis-py#629]
Instead of adding "additional time" to the lock, it should be possible to specify a new timeout for the lock. Consider the following use case:

```
while perform_operation(): # usually completes quickly, but should not take longer than N seconds
    lock.extend(N)
```

In this example, I'd want the lock to expire after no more than N seconds, regardless of how quickly the operation completes. With the current lock implementation this isn't possible. There should be a separate method or argument which indicates that the passed time is the new timeout value (as opposed to being added to the current timeout value).

**Comment by xsh6528** [comment]
I want it too!


    def do_expire(self, expire_time):
        pipe = self.redis.pipeline()
        pipe.watch(self.name)
        lock_value = pipe.get(self.name)
        if lock_value != self.local.token:
            raise LockError("Cannot extend a lock that's no longer owned")
        expiration = pipe.pttl(self.name)
        if expiration is None or expiration < 0:
            # Redis evicted the lock key between the previous get() and now
            # we'll handle this when we call pexpire()
            expiration = 0
        pipe.multi()
        pipe.pexpire(self.name, int(expire_time * 1000))

        try:
            response = pipe.execute()
        except WatchError:
            # someone else acquired the lock
            raise LockError("Cannot extend a lock that's no longer owned")
        if not response[0]:
            # pexpire returns False if the key doesn't exist
            raise LockError("Cannot extend a lock that's no longer owned")
        return True

**Comment by andymccurdy** [comment]
This has been merged in d1279291d68c77709d69fe13aa7ff6d912d98ce4. It will go out to PyPI with the next version.
